File: LeetCode/cpp/UniqueBST.cpp
#include <iostream>
#include <vector>
#include <stack>
#include <queue>

using namespace std;

struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(NULL) {}
    ListNode(int x) : val(x), next(NULL) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(NULL), right(NULL) {}
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    vector<TreeNode*> generate(int start, int end)
    {
        vector<TreeNode*> res;
        for(int val=start; val<=end; val++)
        {
            vector<TreeNode*> left = generate(start, val-1);
            vector<TreeNode*> right = generate(val+1, end);
            for(int i=0; i<left.size(); i++)
            {
                for(int j=0; j<right.size(); j++)
                {
                    TreeNode* node = new TreeNode(val, left[i], right[j]);
                    res.push_back(node);
                }
            }
        }
        if(res.size() == 0)
            res.push_back(nullptr);
        return res;
    }


    vector<TreeNode*> generateTrees(int n)
    {
        vector<TreeNode*> res;
        if(n == 0)
            return res;
        return generate(1, n);
    }

};

File: LeetCode/python/UniqueBST.py
"""
给定n, 返回所有存储1~n的BST.

这道题的解题依据依然是：
当数组为 1，2，3，4，.. i，.. n时，基于以下原则的BST建树具有唯一性：
以i为根节点的树，其左子树由[1, i-1]构成， 其右子树由[i+1, n]构成。 
"""


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.lchild = left
        self.rchild = right

    def is_leaf(self):
        return self.lchild is None and self.rchild is None


class Solution:

    def generateTrees(self, n: int) -> [TreeNode]:

        # EDGE CASE
        if n == 0:
            return []

        # CALL ON  BUILD() TO RETURN THE LIST OF ROOTS
        return self.build(1, n)

    def build(self, start, end):
        trees = []

        # CHOOSE A ROOT VALUE ANYWHERE BETWEEN [START, END]
        for val in range(start, end + 1):

            # ATTACH ROOT TO EVERY POSSIBLE LEFT SUBTREE
            for left in self.build(start, val - 1):

                # ATTACH ROOT TO EVERY POSSIBLE RIGHT SUBTREE
                for right in self.build(val + 1, end):
                    # CREATE A TREE AND ADD IT
                    trees.append(TreeNode(val, left, right))

        # [NONE] = THE EMPTY TREE
        return trees or [None]


if __name__ == "__main__":
    s = Solution()

    res = s.generateTrees(3)
    print(res)
    for r in res:
        print(r.val)
